### PR TITLE
Add [galaxytoolshed] services

### PR DESCRIPTION
**Pull request by guillaume-gricourt**
This PR refers to the PR #7633 and the issue #7660 suggesting one feature, get the number of downloads for a repository, for a new service `galaxytoolshed`.
Many thanks

### CODE REVIEW

File: services/galaxytoolshed/galaxy-toolshed-base.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-base.js:6]: Is it ever possible for this to be an empty array? I suppose this could be analogous to a claimed package name that has no versions/revisions?

If it is possible, then we'll need to account for that empty array scenario in the code that handles the processing of the response. If it's not possible (or at least if we think it should be impossible), then we should update the schema:

```suggestion
const orderedInstallableRevisionsSchema = Joi.array()
  .items(Joi.string().required())
  .required()
```

File: services/galaxytoolshed/galaxy-toolshed-base.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-base.js:29]: Could you explain your thinking around why `schema` is an input parameter?

Something feels off here because there's typically a strict 1:1 relationship between an upstream endpoint we are calling, and the schema we use to validate/make assertions about the structure of the response provided by that upstream endpoint.

Additionally, I assume that because you'd placed this into the base class that it's a response we'll need for other, non-download/non-install related badges?

File: services/galaxytoolshed/galaxy-toolshed-base.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-base.js:45]: My understanding is that we just need to get the revision id/value from the first element in the array (which I'm also assuming will always have at least one element as per above discussion on the schema).

If that's indeed the case, we should just grab the first element instead of grabbing it and modifying the entire collection, i.e.:

```suggestion
      changesetRevision: changesetRevisions[0],
```

the `shift()` function does result in the first element of the array being returned, but it does so by popping that element off the top and then rearranging all the other elements which would be unnecessary and wasteful work (even if it would only have minimal impact in practice)

File: services/galaxytoolshed/galaxy-toolshed-base.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-base.js:55]: I think we can likely get rid of this function entirely, and I'll explain part of why that's the case here as well as on the schema that defines this object which is (currently, but probably shouldn't be) defined in one of the child classes.

Fundamentally I feel like you had this before to aid in some processing which was required based on how you were defining/utilizing the schemas. However, as alluded to in the above comment on the `fetchRepositoryRevisionInstallInfoSchema` definition, we most likely need to change that anyway which should obviate the need for this function.

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-downloads.service.js:10]: As per comments on the base class, this schema definition almost certainly needs to be moved to the base class where the specifics of the associated API are encapsulated.

My impression is that you were trying to use the schemas in more of a class-specific or field-specific manner but that's not really what we want to use the schemas for, and it also adds some unnecessary processing complexity.

Just to reiterate, the schemas define what the response structure should look like for a singular endpoint, and that's still true even in cases like this one where we have multiple badges in a badge family that use different fields from the same response.

What I think we need to do here is to move this schema into the base class to maintain a single definition and encapsulate it alongside its accompanying api call. We can then expand it as needed with additional fields to support additional badges.

Given that the response is structured as a fixed size array with 3 different objects, we need to define our schema to match. At the moment/purely for download count data, that would look something like this:

```suggestion
const repositoryRevisionInstallInfoSchema = Joi.array().ordered(
  Joi.object({
    times_downloaded: Joi.number().required(),
  }).required(),
).items(Joi.any())
```

Note the `.ordered()` call [which guarantees that important ordering](https://joi.dev/api/?v=17.6.0#arrayorderedtype), and the trailing `.items()` which covers the fact that there are other, differently structured elements we don't currently care about. If/when we add more badges that need elements from the second and/or third element from the response, then we can update this schema to match (happy to provide more details when/if we get to that point, but it's basically just adding corresponding entries to the `.ordered()` section

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-downloads.service.js:27]: Should call `renderDownloadsBadge` directly:

```suggestion
      staticPreview: renderDownloadsBadge({ downloads: 10000 }),
```

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-downloads.service.js:38]: We don't need a dedicated `render` function if we're just passing through directly to a render helper like this 

```suggestion
```

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-downloads.service.js:45]: Not strictly necessary, but we can get rid of this too since we just need to grab one field from a fixed, well known position in the response. I'll provide the specifics down below in the `handle` function but if you prefer you could keep this `transform` function and replace this function's body with the suggested line below for extracting the download count

```suggestion
```

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-downloads.service.js:50]: This should go away along with moving the schema to the base class

```suggestion
```

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-downloads.service.js:52]: With the aforementioned changes to the schema, we're guaranteed that `response` will be an array with at least 1 element where the first element will contain the `times_downloaded` field, so we can simply just grab it (this is a large part of why we do response schema validation in the first place, as it greatly simplifies working with/processing the response object)

We can simply destructure that object, and rename the destructured value to `downloads`

```suggestion
    const { times_downloaded: downloads } = response[0]
```

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxy-toolshed-downloads.service.js:53]: Same theme as the `render` function, can call helper directly:

```suggestion
    return renderDownloadsBadge({ downloads })
```

**Review by calebcartwright — COMMENTED** [review]
Thank you for your patience, persistence, and breaking up the original PR as this is much easier to review and discuss!

I think the bones of this are in place, but there's a few changes I'd like to see and a couple questions I've left inline below

File: services/galaxytoolshed/galaxy-toolshed-base.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxy-toolshed-base.js:6]: Yes you're right. This array could be empty due to undefined problems.

File: services/galaxytoolshed/galaxy-toolshed-base.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxy-toolshed-base.js:29]: Exactly, this function will be use by others utilities: 
* releases
* activity
* version
* ...

File: services/galaxytoolshed/galaxy-toolshed-base.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxy-toolshed-base.js:58]: Yes, I forgot this stuff. This is managed by the schema definition. 7a37735b6

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxy-toolshed-downloads.service.js:10]: Ok, I will update this schema for the other utilities. a6c20a5b3

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxy-toolshed-downloads.service.js:27]: Ok, it's modified 059ab9d18

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxy-toolshed-downloads.service.js:50]: Yes, updated into a6c20a5b3

File: services/galaxytoolshed/galaxy-toolshed-downloads.service.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxy-toolshed-downloads.service.js:45]: Updated into 238d9607c

**Comment by calebcartwright** [comment]
Puzzled by the CI failure so changing the title to test some things

**Comment by calebcartwright** [comment]
CI should be fixed. Can you rebase your branch?

File: services/galaxytoolshed/galaxytoolshed-downloads.service.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxytoolshed-downloads.service.js:12]: This is no longer needed

```suggestion
```

File: services/galaxytoolshed/galaxytoolshed-base.js
Comment (calebcartwright) [inline services/galaxytoolshed/galaxytoolshed-base.js:11]: Just want to both confirm that this field will always be present, even when the download count is zero?

Also, we typically wouldn't require this to be changed, but since there's one last round of cleanup needed in the downloads class, this could be changed to utilized our `nonNegativeInteger` helper from the `validators` module (located in the `services` directory)

**Review by calebcartwright — COMMENTED** [review]
Couple minor cleanup items, but happy to approve and merge after they are taken care of!

File: services/galaxytoolshed/galaxytoolshed-downloads.service.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxytoolshed-downloads.service.js]: Remove into 03231b2

File: services/galaxytoolshed/galaxytoolshed-downloads.service.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxytoolshed-downloads.service.js]: Applied into 03231b2

File: services/galaxytoolshed/galaxytoolshed-base.js
Comment (guillaume-gricourt) [reply services/galaxytoolshed/galaxytoolshed-base.js]: Right, the schema is updated with `nonNegativeInteger` 480cc22

**Comment by calebcartwright** [comment]
It's going to be tricky getting this merged because it's coming from an Organization-owned fork. We have a branch protection policy in place that requires the source branch be up to date with the target before PRs can be merged.

We utilize bots and automation which make this seamless for most contributors, but GitHub doesn't allow PRs from Org forks to enable write access to the upstream maintainers (and their bots). This means we'll likely have to try to coordinate a squash/merge update and approval to happen at the same time, as any time any other PR is merged in this repo it'll reset yours back to an unmergable, out-of-date state

**Comment by guillaume-gricourt** [comment]
If I open a new PR, the same as this PR, with a fork coming from my personal account, can be helpful ?

**Comment by calebcartwright** [comment]
If you can go ahead and update this one shortly then we can probably get it resolved. For future PRs though yes it would definitely be worth considering submitting them from a personal fork

**Review by calebcartwright — APPROVED** [review]

**Comment by guillaume-gricourt** [comment]
Thank you so much !

**Pull request merged**
